Shopify/shopify-cli
# Improve `theme serve` errors and add logs for successful operations

This PR introduces two new kinds of log messages for users when they are running a `theme serve` session (**fixed** and **synced**):

<img width="1088" alt="Screenshot 2021-11-24 at 14 46 25" src="https://user-images.githubusercontent.com/1079279/143252709-d40f6a2d-0a58-473f-8a7e-42340bee8376.png">

Now, users get ~a status~ feedback when something happens with a file:
- **Error** - An error has occurred (e.g., a liquid template in an invalid state)
- **Synced** - The local version was successfully uploaded
- **Fixed** - An error was fixed

### WHY are these changes introduced?

These changes address concerns about the error logs raised by the following issues:
- https://github.com/Shopify/shopify-cli/issues/1397 
- https://github.com/Shopify/shopify-cli/issues/1612 

Now, users get feedback (with a **status** and a **timestamp**) when they sync or fix an error in a file.

### WHAT is this pull request doing?

This PR introduces a `ShopifyCLI::Theme::Syncer::StandardReporter` in the syncer class to report non-error messages.

Also, the `ShopifyCLI::Theme::Syncer` has a more generic API now. Instead of `Syncer#delay_errors!` and `Syncer#report_errors!`, we have `Syncer#lock_io!` and `Syncer#unlock_io!` enabling/disabling **reporters**.

So, each reporter is responsible for handling their behavior when they enable/disable.

Finally, the `ShopifyCLI::Theme::Syncer::Operation` is responsible for the multiple versions of operation as a string.


### How to test your changes?

Run the `shopify theme serve`, update some theme files, and check the log of `shopify theme serve`:

![Kapture 2021-11-24 at 14 48 20](https://user-images.githubusercontent.com/1079279/143252749-38f1f9b1-9c69-4359-a4e2-7a64093b1838.gif)

### Update checklist
